File: examples/ospf-kotlin-example/src/main/fuookami/ospf/java/example/core_demo/Demo1.java
package fuookami.ospf.java.example.core_demo;

import java.lang.Integer;
import java.util.ArrayList;
import java.util.Arrays;
import java.util.List;
import java.util.Objects;
import java.util.concurrent.ExecutionException;
import java.util.stream.Collectors;

import fuookami.ospf.kotlin.utils.math.*;
import fuookami.ospf.kotlin.utils.error.Error;
import fuookami.ospf.kotlin.utils.functional.*;
import fuookami.ospf.kotlin.utils.multi_array.*;
import fuookami.ospf.kotlin.core.frontend.variable.*;
import fuookami.ospf.kotlin.core.frontend.expression.monomial.*;
import fuookami.ospf.kotlin.core.frontend.expression.polynomial.*;
import fuookami.ospf.kotlin.core.frontend.expression.symbol.*;
import fuookami.ospf.kotlin.core.frontend.inequality.*;
import fuookami.ospf.kotlin.core.frontend.model.mechanism.*;
import fuookami.ospf.kotlin.core.backend.solver.*;
import fuookami.ospf.kotlin.core.backend.solver.output.*;
import fuookami.ospf.kotlin.core.backend.plugins.scip.*;

public class Demo1 {
    class Company {
        public Integer index;
        public Double capital;
        public Double liability;
        public Double profit;

        public Company(Integer index, Double capital, Double liability, Double profit) {
            this.index = index;
            this.capital = capital;
            this.liability = liability;
            this.profit = profit;
        }
    }

    private final List<Company> companies = Arrays.asList(
            new Company(0, 3.48, 1.28, 5400.0),
            new Company(1, 5.62, 2.53, 2300.0),
            new Company(2, 7.33, 1.02, 4600.0),
            new Company(3, 6.27, 3.55, 3300.0),
            new Company(4, 2.14, 0.53, 900.0)
    );
    private final Double minCapital = 10.0;
    private final Double maxLiability = 5.0;

    private BinVariable1 x;
    private LinearExpressionSymbol capital;
    private LinearExpressionSymbol liability;
    private LinearExpressionSymbol profit;

    private final LinearMetaModel metaModel = new LinearMetaModel();

    public void invoke() {
        metaModel.setName("demo1");
        try {
            initVariable();
            initSymbol();
            initObject();
            initConstraint();
            List<Double> solution = solve();
            if (solution!= null) {
                analyzeSolution(solution);
            }
        } catch (Exception e) {
            e.printStackTrace();
        }
    }

    private void initVariable() {
        x = new BinVariable1("x", new Shape1(this.companies.size()));
        for (Company c : companies) {
            x.get(c.index).setName(String.format("%s_%d", x.getName(), c.index));
        }
        x.forEach(xi -> metaModel.add(xi));
    }

    private void initSymbol() {
        capital = LinearExpressionSymbol.Companion.invoke(
                LinearPolynomialKt.sumLinearMonomials(
                    companies.stream().map(c -> LinearMonomialKt.times(c.capital, x.get(c.index))).collect(Collectors.toList())
                ).toMutable(),
                "capital",
                null
        );
        metaModel.add(capital);

        liability = LinearExpressionSymbol.Companion.invoke(
                LinearPolynomialKt.sumLinearMonomials(
                    companies.stream().map(c -> LinearMonomialKt.times(c.liability, x.get(c.index))).collect(Collectors.toList())
                ).toMutable(),
                "liability",
                null
        );
        metaModel.add(liability);

        profit = LinearExpressionSymbol.Companion.invoke(
                LinearPolynomialKt.sumLinearMonomials(
                    companies.stream().map(c -> LinearMonomialKt.times(c.profit, x.get(c.index))).collect(Collectors.toList())
                ).toMutable(),
                "profit",
                null
        );
        metaModel.add(profit);
    }

    private void initObject() {
        metaModel.maximize(
                profit,
                "profit",
                null
        );
    }

    private void initConstraint() {
        metaModel.addConstraint(
                LinearInequalityKt.geq(capital, minCapital),
                "capital_constraint",
                null,
                false
        );
        metaModel.addConstraint(
                LinearInequalityKt.leq(liability, maxLiability),
                "liability_constraint",
                null,
                false
        );
    }

    private List<Double> solve() throws ExecutionException, InterruptedException {
        LinearSolver solver = new ScipLinearSolver();
        Result<SolverOutput, Error> result = solver.solveAsync(metaModel, null).get();
        if (result != null && result.getOk()) {
            List<Flt64> solution = Objects.requireNonNull(result.getValue()).getSolution();
            metaModel.setSolution(solution);
            return solution.stream().map(Flt64::toDouble).collect(Collectors.toList());
        } else {
            return null;
        }
    }

    private void analyzeSolution(List<Double> solution) {
        ArrayList<Company> result = new ArrayList<>();
        for (Token token : metaModel.getTokens().getTokens()) {
            // if (token.belongsTo(x) && Math.abs(solution.get(token.getSolverIndex()) - 1.0) < 1e-5) {
            if (token.belongsTo(x) && Math.abs(Objects.requireNonNull(token.getDoubleResult()) - 1.0) < 1e-5) {
                int[] vector = token.getVariable().getVectorView();
                result.add(companies.get(vector[0]));
            }
        }
    }
}

File: examples/ospf-kotlin-example/src/main/fuookami/ospf/kotlin/example/core_demo/Demo1.kt
package fuookami.ospf.kotlin.example.core_demo

import fuookami.ospf.kotlin.utils.math.*
import fuookami.ospf.kotlin.utils.concept.*
import fuookami.ospf.kotlin.utils.functional.*
import fuookami.ospf.kotlin.utils.multi_array.*
import fuookami.ospf.kotlin.core.frontend.variable.*
import fuookami.ospf.kotlin.core.frontend.expression.monomial.*
import fuookami.ospf.kotlin.core.frontend.expression.polynomial.*
import fuookami.ospf.kotlin.core.frontend.expression.symbol.*
import fuookami.ospf.kotlin.core.frontend.inequality.*
import fuookami.ospf.kotlin.core.frontend.model.mechanism.*
import fuookami.ospf.kotlin.core.backend.plugins.scip.*

data object Demo1 {
    data class Company(
        val capital: Flt64,
        val liability: Flt64,
        val profit: Flt64
    ) : AutoIndexed(Company::class)

    private val companies: List<Company> = listOf(
        Company(Flt64(3.48), Flt64(1.28), Flt64(5400.0)),
        Company(Flt64(5.62), Flt64(2.53), Flt64(2300.0)),
        Company(Flt64(7.33), Flt64(1.02), Flt64(4600.0)),
        Company(Flt64(6.27), Flt64(3.55), Flt64(3300.0)),
        Company(Flt64(2.14), Flt64(0.53), Flt64(980.0))
    )
    private val minCapital: Flt64 = Flt64(10.0)
    private val maxLiability: Flt64 = Flt64(5.0)

    private lateinit var x: BinVariable1
    private lateinit var capital: LinearExpressionSymbol
    private lateinit var liability: LinearExpressionSymbol
    private lateinit var profit: LinearExpressionSymbol

    private val metaModel: LinearMetaModel = LinearMetaModel("demo1")

    private val subProcesses = listOf(
        Demo1::initVariable,
        Demo1::initSymbol,
        Demo1::initObject,
        Demo1::initConstraint,
        Demo1::solve,
        Demo1::analyzeSolution
    )

    suspend operator fun invoke(): Try {
        for (process in subProcesses) {
            when (val result = process()) {
                is Failed -> {
                    return Failed(result.error)
                }

                else -> {}
            }
        }
        return ok
    }

    private suspend fun initVariable(): Try {
        x = BinVariable1("x", Shape1(companies.size))
        for (c in companies) {
            x[c].name = "${x.name}_${c.index}"
        }
        metaModel.addVars(x)
        return ok
    }

    private suspend fun initSymbol(): Try {
        capital = LinearExpressionSymbol(sum(companies) { it.capital * x[it] }, "capital")
        metaModel.addSymbol(capital)

        liability = LinearExpressionSymbol(sum(companies) { it.liability * x[it] }, "liability")
        metaModel.addSymbol(liability)

        profit = LinearExpressionSymbol(sum(companies) { it.profit * x[it] }, "profit")
        metaModel.addSymbol(profit)
        return ok
    }

    private suspend fun initObject(): Try {
        metaModel.maximize(profit)
        return ok
    }

    private suspend fun initConstraint(): Try {
        metaModel.addConstraint(capital geq minCapital)
        metaModel.addConstraint(liability leq maxLiability)
        return ok
    }

    private suspend fun solve(): Try {
        val solver = SCIPLinearSolver()
        when (val ret = solver(metaModel)) {
            is Ok -> {
                metaModel.tokens.setSolution(ret.value.solution)
            }

            is Failed -> {
                return Failed(ret.error)
            }
        }
        return ok
    }

    private suspend fun analyzeSolution(): Try {
        val ret = ArrayList<Company>()
        for (token in metaModel.tokens.tokens) {
            if (token.result!! eq Flt64.one) {
                ret.add(companies[token.variable.index])
            }
        }
        return ok
    }
}
